staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +190,206 @@ func RecoverFromPanic(err *error) {
+// It tracks errors via the caller stack, the error type and the error message (err.Error() value).
+// An error is considered unique based on these properties (see errKey), with one exception:
+// errors with the same stack and type are considered equal if their message is similar enough (see getVal).
+// To prevent from using an infinite amount of memory, it uses a LRU cache to purge old error values.
+// The cache and count map are separated to allow easy access to all keys, which is required for fuzzy matching.
+type dedupingErrorHandler struct {
+	mutex sync.Mutex
+
+	// cache tracks (stack + type + message) and cleans up old entries in count as they roll off the cache
+	cache *lru.Cache
+	// count tracks (stack + type + message) -> (count + logged)
+	// since rudimentaryErrorBackoff rate limits HandleError to 1000 errors/second,
+	// this counter will effectively never overflow (nothing bad happens even if it does)
+	count map[errKey]errVal
+
+	// errorDepth is how many frames to skip from handleErr
+	errorDepth int
+
+	// delta is the minimum difference between the current time and
+	// errVal.logged required for us to log the associated error again
+	delta time.Duration
+
+	// Similar is the levenshtein ratio used to determine equivalence
+	// TODO unexport this field once we change the default value from 1
+	Similar float64
+
+	// clock allows us to control time in unit tests
+	// TODO add unit tests
+	clock clock.Clock
+}
+
+// errKey tracks uniqueness based on the caller's stack and the type/message of the error
+// it is stored in both the count map and the LRU cache
+// it is removed from the count map when it gets evicted from the LRU cache
+// it is comparable via ==
+type errKey struct {
+	stack   string
+	errType reflect.Type
+	// message is the err.Error() value
+	message string
+}
+
+// errVal tracks how many times we have seen an error, and the last time we logged it
+type errVal struct {
+	count  uint64
+	logged time.Time
+}
+
+// handleErr logs the given error if it is considered new or "not recently logged"
+// currently it logs errors whenever:
+// 1. the associated errKey does not exist in d.count (see d.getVal for specifics on how this is calculated)
+// 2. the associated counter is a power of two
+// 3. the associated logged time's delta from the current time is greater than d.delta
+func (d *dedupingErrorHandler) handleErr(err error) {
+	// the operations below that do not acquire the lock do not mutate d
+
+	// we must determine our stack in this function since getStack counts frames
+	stack := d.getStack()
+	key := errKey{stack: stack, errType: reflect.TypeOf(err), message: err.Error()}
+
+	// operations after this point can mutate d
+	// we must acquire the lock before calling getVal since we cannot
+	// have concurrent reads and writes against d.count
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// increment our counter
+	val, isNewErr := d.getVal(&key)
```

Comment:
is isNewErr needed? Won't new errors have count == 0 (changed to 1 by the next line down)

Reply:
It changes the behavior for the LRU cache.